File: packages/mcp/src/socket-address.ts
import { createHash } from "node:crypto"
import { realpathSync } from "node:fs"
import { tmpdir } from "node:os"
import { join } from "node:path"

export const SOCKET_MODE = 0o600
export const SOCKET_DIR_MODE = 0o700

export type SocketAddressInput = {
  readonly workspace: string
  readonly win?: boolean
  readonly runtimeDir?: string
  readonly tmpdir?: string
  readonly env?: Record<string, string | undefined>
  readonly realpath?: (path: string) => string
}

export const normalizeWorkspacePath = (
  workspace: string,
  win: boolean,
  realpath: (path: string) => string = tryRealpath,
): string => {
  const resolved = realpath(workspace)
  return win ? resolved.toLowerCase() : resolved
}

export const workspaceSocketHash = (
  workspace: string,
  win: boolean,
  realpath?: (path: string) => string,
): string =>
  createHash("sha256").update(normalizeWorkspacePath(workspace, win, realpath ?? tryRealpath))
    .digest("hex")
    .slice(0, 16)

export const socketAddress = (input: SocketAddressInput): string => {
  const win = input.win ?? process.platform === "win32"
  const hash = workspaceSocketHash(input.workspace, win, input.realpath)
  if (win) return `\\\\.\\pipe\\groks-beard-${hash}`
  const env = input.env ?? process.env
  const runtime = input.runtimeDir
    ?? env.XDG_RUNTIME_DIR
    ?? input.tmpdir
    ?? tmpdir()
  return join(runtime, "groks-beard", `${hash}.sock`)
}

const tryRealpath = (path: string): string => {
  try {
    return realpathSync(path)
  } catch {
    return path
  }
}

File: core/src/main/scala/groksbeard/core/SocketAddress.scala
package groksbeard.core

object SocketAddress:
  val SocketMode: Int    = Integer.parseInt("600", 8)
  val SocketDirMode: Int = Integer.parseInt("700", 8)

  def normalize(workspace: String, win: Boolean, realpath: String => String = identity): String =
    val resolved = realpath(workspace)
    if win then resolved.toLowerCase else resolved

  def hash16(normalized: String, sha256Hex: String => String): String =
    sha256Hex(normalized).take(16)

  def address(
      workspace: String,
      win: Boolean,
      sha256Hex: String => String,
      runtimeDir: String,
      realpath: String => String = identity,
  ): String =
    val hash = hash16(normalize(workspace, win, realpath), sha256Hex)
    if win then s"\\\\.\\pipe\\groks-beard-$hash"
    else
      val root = runtimeDir.stripSuffix("/").stripSuffix("\\")
      s"$root/groks-beard/$hash.sock"
  end address
end SocketAddress
